### PR TITLE
add methods to MarcXMLParser

### LINKED ISSUE #333: JabRef Groups interface should work at low resolutions 1366x768
1366x768 - top and bottom bar is there, "OK" button cannot be pressed

### PR DESCRIPTION
**Pull request by Luggas4you**
Add new methods to the class MarcXMLParser, refs https://github.com/JabRef/jabref-issue-melting-pot/issues/101

### ToDo: 
putIssue:
- [x] pages 
- [x] volume and number 
- [x] series and journals stored in different tags 

putThesisDesription: 
- [x] thesis 

putConferenceDetail:
- [x] proceedings (In progress)

### Mandatory checks
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by calixtus** [comment]
Looks good.

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 1]
```diff
@@ -133,8 +136,13 @@ private BibEntry parseEntry(Element element) {
          *  pages
          *  volume and number correct
          *  series and journals stored in different tags
-         *  thesis
-         *  proceedings
+         *  thesis //doctor & masterarbeit (Unterschied?) | type
+         *  proceedings //konferenzbeiträge
+         *  BibTex --> entryType
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:141]:
```suggestion
         *  BibTeX --> entryType
```

Also: Try to write English comments.

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 2]
```diff
@@ -437,6 +447,21 @@ private void putIssue(BibEntry bibEntry, Element datafield) {
         }
     }
 
+    private void putThesisDetails(BibEntry bibEntry, Element datafield) {
+        String thesisDegree = getSubfield("b", datafield);
+        String school = getSubfield("c", datafield);
+
+
+        if ("Masterarbeit".equals(thesisDegree) & StringUtil.isNotBlank(school)) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:455]: Always use && for Boolean.

I think, the condition should be separated. The master Thesis type should be set in all cases

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java (hunk 1 above)
Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:141]: Absolutely, I just made a quick note of them for myself and forgot to delete them...
In the future, I will formulate them neatly right away.

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java (hunk 2 above)
Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:455]: Ok, thank you. 

Even if it is a bachelor's thesis, a habilitation or a dissertation?
I do not need to use the Standard types thesis or PhDThesis at all?

Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:455]: > Ok, thank you. Even if it is a bachelor's thesis, a habilitation or a dissertation?
> 
> Is thesis a general term or is it exclsuive for bachelor's thesis? I guess it makes sense as you rarely need to cite a bachelor's thesis. I update it in my next commit.

Under the tag 502 you will find the type of thesis. A distinction is made here between a dissertation, a master's thesis, a bachelor's thesis and a habilitation thesis. The standard types PhD thesis, master's thesis and thesis are already implemented. Should I then generally mark a thesis entry as master's thesis and a bachelor's thesis a thesis, and a dissertation as a PhD thesis?

**Comment by Luggas4you** [comment]
I looked into proceedings today. However, I could not find a unique identifier. It's often in the title, but I couldn't find a specific tag. 

@calixtus do you happen to know if there is a field that is absolutely required? I couldn't found one :/

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 3]
```diff
@@ -398,6 +404,59 @@ private void putNotes(BibEntry bibEntry, Element datafield) {
...
+        List<String> issues = getSubfields("g", datafield);
+
+        for (String issue : issues) {
+            String[] parts = issue.split(":");
+            if (parts.length == 2) {
+                String key = parts[0].trim();
+                String value = parts[1].trim();
+
+                if (StringUtil.isNotBlank(value)) {
+                    switch (key) {
```
Comment (calixtus) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:419]: Please use modern switch syntax without "break", which is notoriously being forgotten.
case "case" -> { /* statements here */ }

**Comment by calixtus** [comment]
At the very end, it's always a bit of guessing, I think. I believe some fields are not filled anyways if the book is just a conference proceeding.

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java (hunk 3 above)
Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:419]: Fixed

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 4]
```diff
@@ -128,16 +133,17 @@ private BibEntry parseEntry(Element element) {
...
         /*
          * ToDo:
-         *  pages
-         *  volume and number correct
-         *  series and journals stored in different tags
-         *  thesis
-         *  proceedings
+         *  pages //Done
+         *  volume and number correct //Done
+         *  series and journals stored in different tags //Done?
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:138]: Please, don't track TODOs in code. - You can add a markdown checklist to the PR description and then add to the chck boxes.

Example:

```markdown
- [ ] open todo
- [x] done
```

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 5]
```diff
@@ -398,6 +404,45 @@ private void putNotes(BibEntry bibEntry, Element datafield) {
...
+
+    private void putThesisDescription(BibEntry bibEntry, Element datafield) {
+        String thesisDegree = getSubfield("b", datafield);
+        String school = getSubfield("c", datafield);
+        bibEntry.setType(StandardEntryType.MastersThesis);
+
+        if (StringUtil.isNotBlank(school)) {
+            bibEntry.setField(StandardField.SCHOOL, school);
+        }
+        if ("Dissertation".equals(thesisDegree) && StringUtil.isNotBlank(school)) {
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:440]: Isn't the check for `Dissertation` enough? School ist set in line 483, so the setting in 442 can be removed

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 6]
```diff
@@ -406,9 +451,17 @@ private String getSubfield(String a, Element datafield) {
                 return subfield.getTextContent();
             }
         }
+
+
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:455]: Please no two empty lines ^^

Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java]: fixed

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java (hunk 5 above)
Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java]: Fixed (line 438*)

File: src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java [hunk 7]
```diff
@@ -109,10 +110,14 @@ private BibEntry parseEntry(Element element) {
...
             } else if ("490".equals(tag) || "830".equals(tag)) {
                 putSeries(bibEntry, datafield);
+            } else if ("502".equals(tag)) {
+                putThesisDescription(bibEntry, datafield); // Master's thesis, PhD thesis, Thesis
             } else if ("520".equals(tag)) {
                 putSummary(bibEntry, datafield);
             } else if ("653".equals(tag)) {
                 putKeywords(bibEntry, datafield);
+            } else if ("773".equals(tag)) {
+                putIssue(bibEntry, datafield);
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:120]: For the additions, I don't see any test cases (test files!) added. Do you have some at hand to add to the prepository?

Comment (Luggas4you) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:120]: No, I haven't thought about tests yet. I was working on the proceedings. I would like to do the test cases next. There is the BvbFetcherTest class where I can put the tests. Do I write two tests for both methods I have implemented?

Comment (calixtus) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:120]: I would not test the inner code of the fetcher too much, but rather the API and the results. if you put some marcxml with various bibtextypes in it returns the expected bibentry. Would require to build some test files in XML and put to the resources directory in the test directory. There should already be some similar tests.

Comment (calixtus) [reply src/main/java/org/jabref/logic/importer/fileformat/MarcXmlParser.java:120]: Maybe distinguish between the fetcher and the importer. That way you can focus testing the importer if it returns anything usable and test the importer, if he does the parsing right.

**Comment by Luggas4you** [comment]
Implemented the first tests for the methods in MarcXMLParser. I only worked on tests once and therefore have almost no knowledge on them.

**Comment by calixtus** [comment]
Tbh I totally dislike those 100% code coverage tests. Especially here: is it really desirable to have no help page or is this a deficit? 100% coverage is not the target an static code analysis is a trojan horse for bad design.

**Comment by Luggas4you** [comment]
> Tbh I totally dislike those 100% code coverage tests. Especially here: is it really desirable to have no help page or is this a deficit? 100% coverage is not the target and static code analysis is a trojan horse for bad design.

You right, this test does not make any sense. There is no test needed as the GetHelpPage is empty() or not.

**Comment by Luggas4you** [comment]
Why does it say in the "Files changed" section that the BvBFetcher still contains elements of my previous commits even tho I deleted them?

**Comment by Luggas4you** [comment]
This is the original PR https://github.com/JabRef/jabref/pull/10144

**Pull request closed without merging**